### DISCUSSION TITLE
How Can I  Render the Same Handlebar Template for Multiple Data Models

### DISCUSSION THREAD

**Discussion by michael-jehle** [statiqdev/Discussions#229] [Q&A] [answered]
I have one template "page.hbs" which is a handlebar template.
The template uses several handlebar partials.

Thus, my /input folder contains:
- page.hbs
- partial.hbs

Now I need to render this template _multiple times_ with a list of different input data (which is present as a List<MyDataObject>.

My /output folder would thus look like the following:
- page-a.html
- page-b.html
- and so on and so forth

I have seen the EnumerateValues class but could not figure out how to use that along with a List and Handlebar rendering.

Can you help me build a pipeline that would acomplish this task?

Thanks and regards,
Mike

EDIT: this is where I am starting from (1:1 rendering from one .hbs template to one .html file)

    .BuildPipeline("Render page", builder => builder
        .WithInputReadFiles("home.hbs", "Partials/*.hbs")
        .WithProcessModules(
            new RenderHandlebars()
                .WithModel(new MyDataObject(...some data...))
                .WithPartial("head", Config.FromContext(c => c.Inputs.First(i => i.Source.Name == "head.hbs").ContentProvider.GetTextReader().ReadToEnd())))
        .WithOutputWriteFiles(".html"))

**Comment by boblewisvb** [comment, marked as answer]
This is a simplified version of what we are using to develop a static ecommerce site. It uses the following steps.

1. Create module that generates documents with the required data
2. Use module in a pipeline to load data
3. Create pipleline that renders the data into html files

The following should give you enough to create the workflow for your specific situation. Not sure if there is a simpler way, but this is what we have come up with and it is working well.

#### Simple data model
```
public class HandlebarData
{
    public string Name { get; set; }
    public string FileName { get; set; }
}
```

#### 1. Module for creating documents with data
```
public class DocumentsFromHandlebarData : global::Statiq.Common.Module
{
    private readonly IDataService _dataService;

    public DocumentsFromHandlebarData(IDataService dataService)
    {
        _dataService = dataService;
    }

    protected override async Task<IEnumerable<IDocument>> ExecuteContextAsync(IExecutionContext context)
    {
        var data =  await _dataService.GetHandlebarDataAsync(); // returns List<HandlebarData>

        var documents = new List<IDocument>();

        foreach (var model in data)
        {
            var doc = context.CreateDocument(new List<KeyValuePair<string, object>> { new KeyValuePair<string, object>(nameof(HandlebarData), model)});
            documents.Add(doc);
        }

        return documents;
    }
}
```
#### 2. Pipeline for loading data
```
public class LoadHandlebarDataPipeline : Pipeline
{
    public LoadHandlebarDataPipeline(IDataService dataService)
    {
        InputModules = new ModuleList{new DocumentsFromHandlebarData(dataService)};
    }
}
```

#### 3. Pipeline to generate html files
```
public class HandlebarPipeline : Pipeline
{
    public HandlebarPipeline()
    {
        Dependencies.Add(nameof(LoadHandlebarDataPipeline));

        ProcessModules = new ModuleList {

            new ReplaceDocuments(Config.FromContext((ctx) =>
            {
                return ctx.Outputs.FromPipeline(nameof(LoadHandlebarDataPipeline)).AsEnumerable();
            })),

            new MergeContent(new ReadFiles(patterns: "page.hbs")),
            new RenderHandlebars()
                .WithModel(Config.FromDocument((doc)=>doc[nameof(HandlebarData)])),
            new SetDestination(Config.FromDocument((doc, ctx)  => GetDestination(doc))),
        };

        OutputModules = new ModuleList {
            new WriteFiles(),
        };
    }

    private NormalizedPath GetDestination(IDocument doc)
    {
        var data = (HandlebarData)doc[nameof(HandlebarData)];
        return new NormalizedPath($"{data.FileName}.html");
    }
}
```

**Reply by michael-jehle** [reply]
This is very helpful, thanks a lot for your effort!
I was able to introduce this successfully into my project.

**Reply by boblewisvb** [reply]
Glad I could help.  Looking at the code I noticed for the SetDestination the context isn't being used so it could be simplified to 
```
new SetDestination(Config.FromDocument(GetDestination))
```
or even
```
new SetDestination(Config.FromDocument(doc => new NormalizedPath($"{((HandlebarData)doc[nameof(HandlebarData)]).FileName}.html")))
```
